### PR TITLE
Humanoids pr

### PR DESCRIPTION
**Pull request by xavierpuigf**
## Motivation and Context

Adds humanoid control into Habitat. At the moment it adds a kinematically controlled humanoid, and a new action to control their joints and transform. This change is required to support more kinds of agents into Habitat.

## How Has This Been Tested

Passes Pytests

## Types of changes

- Docs change / refactoring / dependency upgrade
- New feature (non-breaking change which adds functionality)

## Checklist

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.
- [X] I have read the **CONTRIBUTING** document.
- [X] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    MobileManipulatorParams,
+)
+
+
+class KinematicHumanoid(MobileManipulator):
+    def _get_humanoid_params(self):
+        return MobileManipulatorParams(
+            arm_joints=[],  # For now we do not add arm_joints
+            gripper_joints=[],
+            wheel_joints=None,
```
Comment (jimmytyyang) [inline habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: Thank you for working on this! Do we expect that humans have wheels?

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+            gripper_open_state=np.array([]),
+            gripper_state_eps=None,
+            wheel_mtr_pos_gain=None,
+            wheel_mtr_vel_gain=None,
+            wheel_mtr_max_impulse=None,
+            ee_offset=[mn.Vector3(), mn.Vector3()],
+            ee_links=[15, 9],
+            ee_constraint=np.zeros((2, 2, 3)),
+            cameras={
+                "robot_head": ArticulatedAgentCameraParams(
```
Comment (jimmytyyang) [inline habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:36]: Thank you for working on this! Do we need another name like "human_head" instead of "robot_head"?

Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:36]: Good point. I think the best would be to replace all these to articulated_agent_agent_head, so that we can get the observations without much code duplication. Maybe this should be in a separate PR?

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 1 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: No, but I found I could not ignore this parameter, even if it was optional. I think that in order to be able to ignore it, we need to set a default argument to wheels, [here](https://github.com/facebookresearch/habitat-lab/blob/main/habitat-lab/habitat/articulated_agents/mobile_manipulator.py#L76) of None. I can do that.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py [hunk 3]
```diff
@@ -467,3 +467,45 @@ def step(self, ee_pos, **kwargs):
...
+                    dtype=np.float32,
+                )
+            }
+        )
+
+    def step(self, **kwargs):
+        new_pos_transform = kwargs["human_joints_trans"]
+        new_joints = new_pos_transform[:-16]
+        new_pos_transform = new_pos_transform[-16:]
+        if np.array(new_pos_transform).sum() != 0:
```
Comment (akshararai) [inline habitat-lab/habitat/tasks/rearrange/actions/actions.py:502]: why is this sum to 0 a good check? Can you add a comment here?

**Review by akshararai — COMMENTED** [review]
Looks good to me. Just needs some additional comments. Should we also add a humanoid URDF and some examples on how to do some random control with the humanoid?

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 1 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: We have [here](https://github.com/facebookresearch/habitat-lab/blob/0cf9e5fae5873d18ed6fde2f93c9694a3bd0f667/habitat-lab/habitat/articulated_agents/articulated_agent_base.py#L83) a check of wheel_joints existing as a param or being `None`, this is why I was proposing to set the default to `None`. If we set it as `False`, or `[]`, we would need to change all these checks as well. I can do that if you prefer.

File: habitat-lab/habitat/articulated_agents/manipulator.py [hunk 4]
```diff
@@ -27,6 +27,7 @@ def __init__(
         limit_robo_joints: bool = True,
         fixed_based: bool = True,
         sim_obj=None,
+        maintain_link_order=False,
```
[the remark this answers is not in the document]
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/manipulator.py:30]: But we may want to have manipulators where this parameter is `True` right?

File: habitat-lab/habitat/articulated_agents/manipulator.py [hunk 5]
```diff
@@ -38,6 +39,7 @@ def __init__(
         self._limit_robo_joints = limit_robo_joints
         self._fixed_base = fixed_based
         self.sim_obj = sim_obj
+        self._maintain_link_order = maintain_link_order
```
[the remark this answers is not in the document]
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/manipulator.py:42]: Same comment as above.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py (hunk 3 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/tasks/rearrange/actions/actions.py:502]: This is a bit hacky... The high level policy calls all the actions at every step, even if those are not executed. When they are not executed, these are simply called as an array of 0s. This works fine in most action spaces since they are relative to the previous agent state, but here the action is absolute. I could make it relative, but given that it involves quaternions and a transformation matrix, it feels like it can complicate things.

**Comment by xavierpuigf** [comment]
> Looks good to me. Just needs some additional comments. Should we also add a humanoid URDF and some examples on how to do some random control with the humanoid?

Good point on the random control, I can add a functions in interactive_play that does that. Regarding the URDF, I made [this PR](https://github.com/facebookresearch/habitat-sim/pull/2016) to download the humanoid URDF in habitat-sim, @aclegg3, is that good to merge?

File: habitat-lab/habitat/articulated_agents/articulated_agent_base.py [hunk 6]
```diff
@@ -27,6 +27,7 @@ def __init__(
         fixed_based: bool = True,
         base_type="mobile",
         sim_obj=None,
+        maintain_link_order=False,
```
Comment (vincentpierre) [inline habitat-lab/habitat/articulated_agents/articulated_agent_base.py:30]: I know there is no docstring for this, but can you add one for this one variable please?

**Comment by xavierpuigf** [comment]
> > Looks good to me. Just needs some additional comments. Should we also add a humanoid URDF and some examples on how to do some random control with the humanoid?
> 
> Good point on the random control, I can add a functions in interactive_play that does that. Regarding the URDF, I made [this PR](https://github.com/facebookresearch/habitat-sim/pull/2016) to download the humanoid URDF in habitat-sim, @aclegg3, is that good to merge?

Alright, I added a random control demo in latest commit

File: examples/interactive_play.py [hunk 7]
```diff
@@ -87,7 +87,12 @@ def step_env(env, action_name, action_args):
 
 
 def get_input_vel_ctlr(
-    skip_pygame, arm_action, env, not_block_input, agent_to_control
+    skip_pygame,
+    arm_action,
+    env,
+    not_block_input,
+    agent_to_control,
+    use_joint_control,
```
Comment (vincentpierre) [inline examples/interactive_play.py:95]: it looks like this only works for humanoids. Is that correct? If so I would rename this argument.

File: examples/interactive_play.py [hunk 8]
```diff
@@ -252,6 +262,42 @@ def get_input_vel_ctlr(
             logger.info("[play.py]: Snapping")
             magic_grasp = 1
 
+    if use_joint_control:
+        # Add random noise to human arms but keep global transform
+        (
+            new_joints,
+            root_trans,
+        ) = env._sim.articulated_agent.get_joint_transform()
+        num_joints = len(new_joints) // 4
```
Comment (vincentpierre) [inline examples/interactive_play.py:271]: where does the 4 come from? Can you create a variable with value 4 with a nice name that describes this magic number?

File: examples/interactive_play.py [hunk 9]
```diff
@@ -612,12 +659,21 @@ def has_pygame():
         action="store_true",
         help="If specified, does not add the inverse kinematics end-effector control.",
     )
+
+    parser.add_argument(
+        "--use-joint-control",
+        action="store_true",
+        default=False,
+        help="Control joints of articulated agent.",
```
Comment (vincentpierre) [inline examples/interactive_play.py:667]: Does this work with any articulated agent? I thought interactive play already supports articulated robots.

File: examples/interactive_play.py [hunk 10]
```diff
@@ -98,10 +103,15 @@ def get_input_vel_ctlr(
     else:
         agent_k = ""
     arm_action_name = f"{agent_k}arm_action"
-    base_action_name = f"{agent_k}base_velocity"
-    arm_key = "arm_action"
-    grip_key = "grip_action"
-    base_key = "base_vel"
+
+    if use_joint_control:
+        base_action_name = f"{agent_k}humanjoint_action"
```
Comment (vincentpierre) [inline examples/interactive_play.py:108]: what happens if `use_joint_control=True` but there is no human in the task ?

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 1 above)
Comment (vincentpierre) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: What is the type of this variable? This should indicate what the no wheel option should be.

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 2 above)
Comment (vincentpierre) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:36]: I agree a separate PR would help us move fast on this. Please make a note this will need to be changed later.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py [hunk 11]
```diff
@@ -467,3 +467,48 @@ def step(self, ee_pos, **kwargs):
...
+                    shape=(4 * (num_joints + num_dim_transform),),
+                    low=-1,
+                    high=1,
+                    dtype=np.float32,
+                )
+            }
+        )
+
+    def step(self, **kwargs):
+        new_pos_transform = kwargs["human_joints_trans"]
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/actions/actions.py:499]: Can you just change the signature of step to have human_joints_trans argument? This looks too hacky.

Comment (vincentpierre) [reply habitat-lab/habitat/tasks/rearrange/actions/actions.py:499]: This will also allow you to add a docstring explaining the all zeros trick used here.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py [hunk 12]
```diff
@@ -467,3 +467,48 @@ def step(self, ee_pos, **kwargs):
...
+        if np.array(new_pos_transform).sum() != 0:
+            vecs = [
+                mn.Vector4(new_pos_transform[i * 4 : (i + 1) * 4])
+                for i in range(4)
+            ]
+            new_transform = mn.Matrix4(*vecs)
+            self.cur_articulated_agent.set_joint_transform(
+                new_joints, new_transform
+            )
+        return self._sim.step(HabitatSimActions.changejoint_action)
```
Comment (vincentpierre) [inline habitat-lab/habitat/tasks/rearrange/actions/actions.py:514]: I do not understand what this does. Why is this needed? What does self._sim need to update for the changejoint_action ?

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 1 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: `Optional[List[int]]`

File: examples/interactive_play.py (hunk 9 above)
Comment (xavierpuigf) [reply examples/interactive_play.py:667]: It doesn't, I could instead have a single flag for human, or check if human is in the config file to activate this.

File: examples/interactive_play.py (hunk 10 above)
Comment (xavierpuigf) [reply examples/interactive_play.py:108]: Same comment as above.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py (hunk 12 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/tasks/rearrange/actions/actions.py:514]: I probably misunderstood the code, but don't we need to call `sim.step` on every step() function, since functions calling step expect observations from the environment? I think this makes sense when we call this function via the high level policy.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py (hunk 11 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/tasks/rearrange/actions/actions.py]: Addressed in latest commit

File: examples/interactive_play.py (hunk 8 above)
Comment (xavierpuigf) [reply examples/interactive_play.py:271]: A variable is a bit overkill since it corresponds to 4 dimensions per quaternion, and is used in the subsequent indexing. I added a comment right there.

File: examples/interactive_play.py (hunk 7 above)
Comment (xavierpuigf) [reply examples/interactive_play.py:95]: Resolved in last commit

File: examples/interactive_play.py (hunk 8 above)
Comment (akshararai) [reply examples/interactive_play.py:273]: Hm I don't see this comment. Are the latest changes pushed to the branch?

File: examples/interactive_play.py (hunk 7 above)
Comment (akshararai) [reply examples/interactive_play.py]: Hm I still see the variable name as `use_joint_control `. Is the change pushed?

File: examples/interactive_play.py [hunk 13]
```diff
@@ -252,6 +262,42 @@ def get_input_vel_ctlr(
             logger.info("[play.py]: Snapping")
             magic_grasp = 1
 
+    if use_joint_control:
+        # Add random noise to human arms but keep global transform
+        (
+            new_joints,
```
Comment (akshararai) [inline examples/interactive_play.py:268]: I think `new_joints` is a bad variable name. Should we call it `joint_trans` to match `root_trans` ?

File: examples/interactive_play.py [hunk 14]
```diff
@@ -252,6 +262,42 @@ def get_input_vel_ctlr(
...
+        ]
+        rotated_joints_quat = []
+        for index, joint_quat in enumerate(new_joints_quat):
+            random_vec = np.random.rand(3)
+            random_angle = np.random.rand() * 360 / 60
+            rotation_quat = mn.Quaternion.rotation(
+                mn.Rad(random_angle), mn.Vector3(random_vec).normalized()
+            )
+            new_quat = joint_quat
+            if index > 10:
```
Comment (akshararai) [inline examples/interactive_play.py]: why only for index > 10?

File: examples/interactive_play.py [hunk 15]
```diff
@@ -252,6 +262,42 @@ def get_input_vel_ctlr(
...
+            for index in range(num_joints)
+        ]
+        rotated_joints_quat = []
+        for index, joint_quat in enumerate(new_joints_quat):
+            random_vec = np.random.rand(3)
+            random_angle = np.random.rand() * 360 / 60
+            rotation_quat = mn.Quaternion.rotation(
+                mn.Rad(random_angle), mn.Vector3(random_vec).normalized()
+            )
+            new_quat = joint_quat
```
Comment (akshararai) [inline examples/interactive_play.py:291]:
```suggestion
```

File: examples/interactive_play.py [hunk 16]
```diff
@@ -252,6 +262,42 @@ def get_input_vel_ctlr(
...
+            if index > 10:
+                new_quat = joint_quat * rotation_quat
```
Comment (akshararai) [inline examples/interactive_play.py:293]:
```suggestion
                joint_quat *= rotation_quat
```

File: examples/interactive_play.py [hunk 17]
```diff
@@ -252,6 +262,42 @@ def get_input_vel_ctlr(
...
+            rotated_joints_quat.append(new_quat)
```
Comment (akshararai) [inline examples/interactive_play.py:294]:
```suggestion
            rotated_joints_quat.append(joint_quat)
```

File: examples/interactive_play.py (hunk 9 above)
Comment (akshararai) [reply examples/interactive_play.py]: I would make this an explicit flag like `use-humanoid-agent`. I think we should revamp the `interactive_play.py` script in the future, so just use the simplest, clearest option.

File: examples/interactive_play.py [hunk 18]
```diff
@@ -612,12 +659,21 @@ def has_pygame():
...
+        help="Control joints of articulated agent.",
+    )
+
     parser.add_argument(
         "--gfx",
         action="store_true",
         default=False,
         help="Save a GFX replay file.",
     )
+
```
Comment (akshararai) [inline examples/interactive_play.py:680]:
```suggestion
```

**Review by akshararai — COMMENTED** [review]
Looks good to me. Once all comments are addressed, we can merge this soon I think. Thank you for working on this, and addressing all the comments!

File: examples/interactive_play.py (hunk 9 above)
Comment (xavierpuigf) [reply examples/interactive_play.py]: Resolved

File: habitat-lab/habitat/articulated_agents/mobile_manipulator.py [hunk 19]
```diff
@@ -112,6 +112,7 @@ def __init__(
         sim: Simulator,
         limit_robo_joints: bool = True,
         fixed_base: bool = True,
+        maintain_link_order: bool = True,
```
[the remark this answers is not in the document]
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/mobile_manipulator.py]: We want to be able to have mobile manipulators where we will set this variable

File: habitat-lab/habitat/articulated_agents/mobile_manipulator.py [hunk 20]
```diff
@@ -141,6 +142,7 @@ def __init__(
             fixed_based=fixed_base,
             sim_obj=self.sim_obj,
             base_type=base_type,
+            maintain_link_order=maintain_link_order,
```
[the remark this answers is not in the document]
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/mobile_manipulator.py:148]: We want to be able to have mobile manipulators where we will set this variable

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py [hunk 21]
```diff
@@ -0,0 +1,80 @@
...
+                    cam_offset_pos=mn.Vector3(0.0, 0.5, 0.25),
+                    cam_look_at_pos=mn.Vector3(0.0, 0.5, 0.75),
+                    attached_link_id=-1,
+                ),
+                "robot_third": ArticulatedAgentCameraParams(
```
[the remark this answers is not in the document]
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:41]: Will create a PR soon for this

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 2 above)
Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:36]: Note taken

File: examples/interactive_play.py (hunk 13 above)
Comment (xavierpuigf) [reply examples/interactive_play.py:268]: Should be there now

File: examples/interactive_play.py (hunk 17 above)
Comment (xavierpuigf) [reply examples/interactive_play.py]: Updated

File: examples/interactive_play.py (hunk 14 above)
Comment (xavierpuigf) [reply examples/interactive_play.py]: Added a comment to explain. Its just to make the demo look nicer, no particular reason.

File: habitat-lab/habitat/articulated_agents/manipulator.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (akshararai) [reply habitat-lab/habitat/articulated_agents/manipulator.py:30]: sure, makes sense. I think the order is only ambiguous for parallel chains, like legs, two arms. For a single chain like a manipulator this order is actually fixed. But good to have this flag here to make it explicit.

**Review by akshararai — APPROVED** [review]
looks good to me.

File: habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py (hunk 1 above)
Comment (jimmytyyang) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: yeah, I faced the same issue before, my solution at that time was to use inheritance, which is not the best way to do this

Comment (xavierpuigf) [reply habitat-lab/habitat/articulated_agents/humanoids/kinematic_humanoid.py:23]: Could you elaborate on how you used inheritance? I think the best would be to make it List[int] (no optional) with a default argument `[]` and whenever we are checking this param, we replace the `None/False` checks fro empty list.

File: habitat-lab/habitat/tasks/rearrange/actions/actions.py (hunk 12 above)
Comment (ASzot) [reply habitat-lab/habitat/tasks/rearrange/actions/actions.py:520]: This could cause the simulator to step twice per high-level step. The action should only step the simulator if the action is the last action to execute. See how it's [done in the other actions](https://github.com/facebookresearch/habitat-lab/blob/2ca5b88bf86bb9fef790eebee9dddc08bc0bf4c9/habitat-lab/habitat/tasks/rearrange/actions/actions.py#L403).
